vllm/model_executor/models/h2ovl.py

```diff
@@ -0,0 +1,400 @@
+        pixel_values2 = load_image2(
+            image,
+            input_size=input_size,
+            min_num=min_num,
+            max_num=max_num,
+            target_aspect_ratio=target_aspect_ratio,
+        )
+        pixel_values = torch.cat(
+            [pixel_values2[:-1], pixel_values[:-1], pixel_values2[-1:]], 0)
+
+    else:
+        transform = build_transform(input_size=input_size)
+        images, _ = dynamic_preprocess(
+            image,
+            min_num=min_num,
+            max_num=max_num,
+            image_size=input_size,
+            use_thumbnail=use_thumbnail,
+        )
+        pixel_values = [transform(image) for image in images]
+        pixel_values = torch.stack(pixel_values)
+
+    return pixel_values
+
+
+def image_to_pixel_values_wrapper(hf_config: PretrainedConfig,
+                                  max_dynamic_patch: Optional[int] = None):
+    image_size = hf_config.vision_config.image_size
+    min_num = hf_config.min_dynamic_patch
+    if max_dynamic_patch is None:
+        max_dynamic_patch = hf_config.max_dynamic_patch
+    use_thumbnail = hf_config.use_thumbnail
+    use_MSAC = hf_config.use_msac
+    return partial(
+        image_to_pixel_values,
+        input_size=image_size,
+        min_num=min_num,
+        max_num=max_dynamic_patch,
+        use_thumbnail=use_thumbnail,
+        use_MSAC=use_MSAC,
+    )
+
+
+def get_max_internvl_image_tokens(ctx: InputContext,
+                                  *,
+                                  max_dynamic_patch: Optional[int] = None):
+    """
+    Calculate the maximum number of tokens with/without MSAC and thumbnail
+    """
+    hf_config = ctx.get_hf_config()
+    use_thumbnail = hf_config.use_thumbnail
+    use_MSAC = hf_config.use_msac
+
+    if max_dynamic_patch is None:
+        max_dynamic_patch = hf_config.max_dynamic_patch
+
+    num_patches = get_internvl_num_patches(hf_config)
+
+    coefficient = 2 if use_MSAC else 1
+    num_blocks = coefficient * max_dynamic_patch + (1 if use_thumbnail else 0)
+
+    return num_blocks * num_patches
+
+
+class H2OVLInputPipeline(InternVLInputPipeline):
+
+    def __init__(self):
+        super().__init__(IMG_START, IMG_END, IMG_CONTEXT)
+
+    def input_processor(
+        self,
+        ctx: InputContext,
+        inputs: DecoderOnlyInputs,
+        *,
+        max_dynamic_patch: Optional[int] = None,
+    ) -> DecoderOnlyInputs:
+        multi_modal_data = inputs.get("multi_modal_data")
+        if multi_modal_data is None or "image" not in multi_modal_data:
+            return inputs
+
+        model_config = ctx.model_config
+        hf_config = ctx.get_hf_config()
+
+        image_data = multi_modal_data["image"]
+        num_patches = get_internvl_num_patches(hf_config)
+
+        # can only get the total blocks num after the image fully processed
+        num_blocks_calculator = image_to_pixel_values_wrapper(
+            hf_config, max_dynamic_patch=max_dynamic_patch)
```

Comment:
We should definitely avoid calling the image processing twice. If there's no way to get around this, you could also return the processed pixel values within `input_processor`, and make `input_mapper` a no-op function. 

Molmo has done something similar here https://github.com/vllm-project/vllm/blob/eae3d48181b1ad27f132f14df18e8cff203f7552/vllm/model_executor/models/molmo.py#L860-L866

Reply:
Thanks for the suggestion, I followed the approach in molmo.py to prepare the `multi_modal_data` within `input_processor`, so there’s no processing for real image data in input_mapper. I’ve kept the image processing code for dummy_data, as I intend to reuse the dummy_data from InternVL.